### PR TITLE
[Local GC] Move knowledge of the free object method table outside of the GC

### PR DESCRIPTION
**Pull request by swgillespie**
The EE provides a method table for a specific type of object, the "free object", which is used by the GC to keep the heap crawlable. A free object is conceptually similar to a managed array in that it consists of an object header and a field indicating how many array elements the object has (i.e. how large the free region is), followed by a region of space known to be free to the GC.

This PR ensures that the GC's knowledge of the free method table is acquired through `GCToEEInterface` upon startup. This PR also has the GC acquire two additional details on startup: 1) the size of `ArrayBase`, and the 2) offset of the `NumComponents` field of `ArrayBase`. These are used to locate the offset of the field into which the size of the free object is stored and to locate the first byte of the free object's free region.

### CODE REVIEW

File: src/gc/env/gcenv.ee.h [hunk 1]
```diff
@@ -71,6 +71,9 @@ class GCToEEInterface
 
     static void HandleFatalError(unsigned int exitCode);
     static bool ShouldFinalizeObjectForUnload(AppDomain* pDomain, Object* obj);
+    static MethodTable* GetFreeObjectMethodTable();
+    static size_t GetSizeOfArrayBase();
```
Comment (jkotas) [inline src/gc/env/gcenv.ee.h]: These constants are used on very perf sensitive paths. I do not think it is ok to virtualize them. They have to be part of the fixed binary contract between VM and GC.

Comment (swgillespie) [inline src/gc/env/gcenv.ee.h]: These methods are called once on GC startup and cached on the EE side - is that still a problem?

Comment (jkotas) [inline src/gc/env/gcenv.ee.h]: Yes.

Comment (swgillespie) [inline src/gc/env/gcenv.ee.h]: Ok, will fix!

**Comment by swgillespie** [comment]
cc @adityamandaleeka @Maoni0 @sergiy-k

**Comment by jkotas** [comment]
Need to resolve conflicts ... LGTM otherwise.

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**